nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java

```diff
@@ -0,0 +1,268 @@
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        JsonNode firstJsonNode;
+        try (final InputStream in = session.read(original)) {
+            firstJsonNode = readJson(in);
+        } catch (final Exception e) {
+            logger.error("Failed to transform {}; routing to failure", original, e);
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        final String jsonString;
+        final String transform;
+        try {
+            transform = context.getProperty(JSLT_TRANSFORM).evaluateAttributeExpressions(original).getValue();
+            Expression jsltExpression = transformCache.get(transform, currString -> Parser.compileString(transform));
+
+            final JsonNode transformedJson = jsltExpression.apply(firstJsonNode);
+            if (transformedJson == null) {
+                jsonString = "";
+                logger.warn("JSLT transform resulted in no data");
+            } else {
+                jsonString = context.getProperty(PRETTY_PRINT).asBoolean() ? transformedJson.toPrettyString() : transformedJson.toString();
+            }
+        } catch (final Exception ex) {
+            logger.error("Unable to transform {} due to {}", new Object[]{original, ex.toString()}, ex);
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        FlowFile transformed = session.write(original, out -> out.write(jsonString.getBytes(StandardCharsets.UTF_8)));
+
+        transformed = session.putAttribute(transformed, CoreAttributes.MIME_TYPE.key(), "application/json");
+        session.transfer(transformed, REL_SUCCESS);
+        session.getProvenanceReporter().modifyContent(transformed, "Modified With " + transform, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+        logger.debug("Transformed {}", original);
+    }
+
+    @OnStopped
+    @OnShutdown
+    public void onStopped(ProcessContext context) {
+        transformCache.cleanUp();
+    }
+
+    /**
+     * Takes an InputStream and returns the top-level JsonNode returned for the parsed JSON object
+     *
+     * @param in the InputStream from which to read the input
+     * @return the top-level JsonNode for the parsed JSON object
+     * @throws IOException if an error occurs while parsing JSON from the InputStream
+     */
+    private JsonNode readJson(final InputStream in) throws IOException {
+        JsonNode firstJsonNode;
+        try {
+            JsonParser jsonParser = jsonFactory.createParser(in);
+            jsonParser.setCodec(codec);
+
+            JsonToken token = jsonParser.nextToken();
+            if (token == JsonToken.START_ARRAY) {
+                token = jsonParser.nextToken(); // advance to START_OBJECT token
+            }
```

Comment:
This approach only reads the first element in an array. It seems like this could be confusing behavior, should the Processor require all incoming JSON to be an Object at the root level? If JSLT does not support transforming array root elements, then it seems better to enforce a standard. It may be worth noting this detail in the description one way or the other.

Reply:
Great catch! I want it to support array root elements like JoltTransformJSON does. One difference between JoltTransformJSON and JoltTransformRecord is that for the former you write the transform for the entire input (so if you want to iterate over records you need to put that `*` in the spec, and for the latter you write the transform for each object (record). I want to have the same behavior for JSLTTransformJSON, and someday would like to add a JSLTTransformRecord in the same vein. Will update this (and the above comment) to apply the transform to the entire input and add a unit test with a top-level array.